src/core/pinned_memory_manager.cc

```diff
@@ -0,0 +1,226 @@
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <sstream>
+
+#include "src/core/logging.h"
+#include "src/core/pinned_memory_manager.h"
+
+#ifdef TRTIS_ENABLE_GPU
+#include <cuda_runtime_api.h>
+#endif  // TRTIS_ENABLE_GPU
+
+namespace nvidia { namespace inferenceserver {
+
+namespace {
+
+std::string
+PointerToString(void* ptr)
+{
+  std::stringstream ss;
+  ss << ptr;
+  return ss.str();
+}
+
+}  // namespace
+
+std::unique_ptr<PinnedMemoryManager> PinnedMemoryManager::instance_;
+
+Status
+PinnedMemoryManager::Create(const Options& options)
+{
+  if (instance_ != nullptr) {
+    return Status(
+        RequestStatusCode::ALREADY_EXISTS,
+        "PinnedMemoryManager has been created");
+  }
+
+  instance_.reset(new PinnedMemoryManager(options));
+  return Status::Success;
+}
+
+Status
+PinnedMemoryManager::Alloc(
+    void** ptr, size_t size, bool allow_nonpinned_fallback)
+{
+  if (instance_ == nullptr) {
+    return Status(
+        RequestStatusCode::UNAVAILABLE,
+        "PinnedMemoryManager has not been created");
+  }
+
+  auto status = Status::Success;
+  {
+    // only holds the lock on accessing manager member
+    std::lock_guard<std::mutex> lk(instance_->mtx_);
+
+    status = instance_->CheckPrerequisite(size);
+    // treat as if the operation will succeed to avoid over-subscription
+    instance_->total_pinned_byte_size_ += size;
+  }
+
+  // allocate buffer
+  bool is_pinned = true;
+  if (status.IsOk()) {
+#ifdef TRTIS_ENABLE_GPU
+    auto err = cudaHostAlloc(ptr, size, cudaHostAllocPortable);
+    if (err != cudaSuccess) {
+      // set to nullptr on error to avoid freeing invalid pointer
+      *ptr = nullptr;
+      status = Status(
+          RequestStatusCode::INTERNAL,
+          "failed to allocate pinned system memory: " +
+              std::string(cudaGetErrorString(err)));
+    }
+#else
+    *ptr = nullptr;
+    status = Status(
+        RequestStatusCode::INTERNAL,
+        "failed to allocate pinned system memory: " +
+            "TRTIS_ENABLE_GPU is not set"));
+#endif  // TRTIS_ENABLE_GPU
+  }
+  if ((!status.IsOk()) && allow_nonpinned_fallback) {
+    static bool warning_logged = false;
+    if (!warning_logged) {
+      LOG_WARNING << status.Message()
+                  << ", allocating non-pinned system memory";
```

Comment:
"... pinned memory pool exceeded, falling back to non-pinned system memory".

Reply:
The reason for not allocating pinned memory will be explained in `status.Message()`, so the new commit only changes here to ", falling back to non-pinned system memory"